pkg/observable/channel/observer.go

```diff
@@ -0,0 +1,125 @@
+	// observer from observers in a concurrency-safe manner.
+	onUnsubscribe func(toRemove *channelObserver[V])
+	// observerMu protects the observerCh and closed fields.
+	observerMu *sync.RWMutex
+	// observerCh is the channel that is used to emit values to the observer.
+	// I.e. on the "N" side of the 1:N relationship between observable and
+	// observer.
+	observerCh chan V
+	// closed indicates whether the observer has been closed. It's set in
+	// unsubscribe; closed observers can't be reused.
+	closed bool
+}
+
+type UnsubscribeFunc[V any] func(toRemove *channelObserver[V])
+
+func NewObserver[V any](
+	ctx context.Context,
+	onUnsubscribe UnsubscribeFunc[V],
+) *channelObserver[V] {
+	// Create a channel for the observer and append it to the observers list
+	return &channelObserver[V]{
+		ctx:           ctx,
+		observerMu:    new(sync.RWMutex),
+		observerCh:    make(chan V, observerBufferSize),
+		onUnsubscribe: onUnsubscribe,
+	}
+}
+
+// Unsubscribe closes the subscription channel and removes the subscription from
+// the observable.
+func (obsvr *channelObserver[V]) Unsubscribe() {
+	obsvr.unsubscribe()
+}
+
+// Ch returns a receive-only subscription channel.
+func (obsvr *channelObserver[V]) Ch() <-chan V {
+	return obsvr.observerCh
+}
+
+// unsubscribe closes the subscription channel, marks the observer as closed, and
+// removes the subscription from its observable's observers list via onUnsubscribe.
+func (obsvr *channelObserver[V]) unsubscribe() {
+	obsvr.observerMu.Lock()
+	defer obsvr.observerMu.Unlock()
+
+	if obsvr.closed {
+		return
+	}
+
+	close(obsvr.observerCh)
+	obsvr.closed = true
+	obsvr.onUnsubscribe(obsvr)
+}
+
+// notify is used called by observable to send on the observer channel. Can't
+// use channelObserver#Ch because it's receive-only. It will block if the channel
+// is full but will release the read-lock for half of the sendRetryInterval. The
+// other half holds is spent holding the read-lock and waiting for the (full)
+// channel to be ready to receive.
+func (obsvr *channelObserver[V]) notify(value V) {
+	defer obsvr.observerMu.RUnlock()
+
+	// if observerCh is full, release the lock and try again every sendRetryInterval.
+	sendRetryTicker := time.NewTicker(sendRetryInterval / 2)
+	for {
+		// observerMu must remain read-locked until the value is sent on observerCh
+		if !obsvr.observerMu.TryRLock() {
+			time.Sleep(sendRetryInterval / 2)
+			continue
+		}
+		if obsvr.closed {
+			return
+		}
+
+		select {
+		case <-obsvr.ctx.Done():
+			// if the context is done just release the read-lock (deferred)
+			return
+		case obsvr.observerCh <- value:
+			return
+		case <-sendRetryTicker.C:
+			// CONSIDERATION: repurpose this retry loop into a default path which
+			// buffers values so that the producer doesn't block and other observers
+			// can still be notified.
+			// TECHDEBT: add some logic to drain the buffer at some appropriate time
```

Comment:
> Are we sure we want to do this? It would mean this is an "at most once fan-out service"

I'm not sure I followed what you mean by "at most once". Are you referring to the number of subscribers which can receive a given value (A), the number of times any subscriber can receive a given value (B), or something else? 

A) Draining here would only affect this observer's (a single, slow, subscriber's) view of the stream of values. 

B) Subscribers are already only able to read a given value once as once it's received from the channel (regardless of that channel's buffer size).